### PR TITLE
feat: Finalize the SLSA v1.0 provenance format

### PR DESCRIPTION
**Pull request by asraa**
**Fixes issue:**
Fixes https://github.com/in-toto/in-toto-golang/issues/197

**Description:**
Finalizes the SLSA v1.0 provenance format.

@adityasaky it would be great to merge the `next` branch into main soon and do a minor release with the SLSA v1 feature!

**Please verify and check that the pull request fulfills the following
requirements:**

- [ x ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

cc @marklodato @laurentsimon @chuangw6

### CODE REVIEW

**Review by chuangw6 — COMMENTED** [review]
Thank you @asraa !!
Can we add a provenance statement struct in model.go like [these](https://github.com/in-toto/in-toto-golang/blob/a813798404c5f3d9ed27ec14ef4915f7e913b375/in_toto/model.go#L993-L1010)?

**Comment by chuangw6** [comment]
Also want to loop in @sherzberg-1 as fyi. Just want to make sure the intoto provenance statement can be serialized and deserialized to grafeas proto.

**Comment by sherzberg-1** [comment]
Thanks @chuangw6 , will make sure they can serialize/deserialize back and forth.

**Comment by MarkLodato** [comment]
LGTM

**Review by MarkLodato — APPROVED** [review]

**Comment by chuangw6** [comment]
Hi @adityasaky ,
[SLSA v1 is released](https://openssf.org/press-release/2023/04/19/openssf-announces-slsa-version-1-0-release/). And I think this PR is up to date.
Can you please TAL at this PR and see if we can get the branch into main and cut a release? 
Thanks!

File: in_toto/slsa_provenance/v1.0/provenance.go
Comment (adityasaky) [inline in_toto/slsa_provenance/v1.0/provenance.go:95]: Thoughts on using ResourceDescriptor from https://github.com/in-toto/attestation/blob/main/go/v1/resource_descriptor.pb.go?

cc @marcelamelara

**Review by adityasaky — COMMENTED** [review]
Thanks @asraa! I think we can merge this overall.

However, I do want to hear from @in-toto/attestation-maintainers re using the models generated for ResourceDescriptor (and possibly Statement etc as well).

File: in_toto/slsa_provenance/v1.0/provenance.go
Comment (chuangw6) [reply in_toto/slsa_provenance/v1.0/provenance.go:95]: I think that's a good idea. But given the proto only defines ResourceDescriptor and misses many other fields of SLSA v1 predicate, should we move on with current approach and cut a release first, and later switch over to models generated from proto once all v1 predicate fields are well defined in attestation repo?

Comment (marcelamelara) [reply in_toto/slsa_provenance/v1.0/provenance.go:95]: To my understanding, SLSA already defines full protos for SLSA provenance v1: https://github.com/slsa-framework/slsa/tree/main/docs/provenance/schema/v1

We (in-toto attestation maintainers) were operating under the assumption that SLSA will maintain all SLSA-related predicate spec and proto versions. Based on your comment here, it sounds like there's an expectation that the in-toto attestation repo should maintain (a separate copy) of SLSA provenance v1 protos?

Comment (chuangw6) [reply in_toto/slsa_provenance/v1.0/provenance.go:95]: Oh, I was only looking at https://github.com/in-toto/attestation/tree/main/protos/in_toto_attestation/v1. I thought intoto will define its own proto again.

If intoto can reuse the proto from SLSA, that would be great. I have no option on this :D I'd defer intoto maintainers to decide this.

I am looking forward to seeing an intoto-golang release for SLSA v1!

Comment (marcelamelara) [reply in_toto/slsa_provenance/v1.0/provenance.go:95]: Got it :) At least my preference would be to use the SLSA-defined protos for generating the language bindings for the predicate, so this integration is something we should definitely discuss on our end.

Comment (chuangw6) [reply in_toto/slsa_provenance/v1.0/provenance.go:95]: Sounds good to me! Is there a target date we can expect a release with the slsa v1 support? Thanks!

File: in_toto/slsa_provenance/v1.0/provenance.go
Comment (adityasaky) [inline in_toto/slsa_provenance/v1.0/provenance.go:11]: Since the predicate type will not include the minor version, can this package dir be just `v1` instead of `v1.0`?

> Minor version changes are always backwards compatible and “monotonic.” Such changes do not update the predicateType.

https://slsa.dev/provenance/v1#parsing-rules

File: in_toto/slsa_provenance/v1.0/provenance.go
Comment (adityasaky) [reply in_toto/slsa_provenance/v1.0/provenance.go:95]: I think I'm okay to merge this PR and then next into main and cut a release to unblock other tools. My inclination is to switch it out behind the scenes later as we want to do that for multiple models, not just SLSA Provenance.

**Review by adityasaky — APPROVED** [review]
LGTM, I do have a question about the package directory.

File: in_toto/slsa_provenance/v1.0/provenance.go
Comment (pxp928) [reply in_toto/slsa_provenance/v1.0/provenance.go:95]: +1 with @adityasaky can merge for now and switch behind the scenes in another release.

**Review by pxp928 — APPROVED** [review]
LGTM

File: in_toto/slsa_provenance/v1.0/provenance.go
Comment (chuangw6) [reply in_toto/slsa_provenance/v1.0/provenance.go:11]: > My thinking is that if Provenance v1.1 introduces a change, it'd go in in_toto/slsa_provenance/v1/provenance.go. I'm wary of having in_toto/slsa_provenance/v1.0/provenance.go, in_toto/slsa_provenance/v1.1/provenance.go, etc. Since they're backwards compatible from here, I see no reason why we can't have a single v1/provenance.go.

SGTM and agreed!

**Pull request merged**
